### PR TITLE
Initial pass at tree interfaces

### PR DESCRIPTION
**Pull request by DLehenbauer**

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,39 @@
+ARG VARIANT=12
```
Comment (DLehenbauer) [inline .devcontainer/Dockerfile:1]: I'll separate out the .devcontainer changes and commit separately.  This one still needs a bit of work.

If you're not familiar: https://code.visualstudio.com/docs/remote/containers

Easiest way to try it out:
* Install WSL2
* Install Docker Desktop
* Install "Remote: Containers" extension
* Use "> Remote-Containers: Open repository in remote container..."

Viola...  preconfigured dev environment that is 100% isolated from your dev machine, but still "feels local".

File: packages/core/tree/bench/index.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { run } from "hotloop";
```
Comment (DLehenbauer) [inline packages/core/tree/bench/index.ts:1]: Yeah, these aren't tree benchmarks...  I should write some.  :-)

File: packages/core/tree/src/consumerset.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
+export type ConsumerSet<T> = undefined | T | Array<{ value: T, count: number }>
```
Comment (DLehenbauer) [inline packages/core/tree/src/consumerset.ts:1]: Duped code.  Should refactor.

File: packages/core/tree/src/handletable.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
+/*!
```
Comment (DLehenbauer) [inline packages/core/tree/src/handletable.ts:1]: Duped from FF.  Should move someplace shared.

File: packages/core/tree/test/bottomuptree.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    protected isDirty(node: TreeNode) {
+        return this.dirty[node] !== false;
+    }
+
+    protected clearDirty(node: TreeNode) {
+        this.dirty[node] = false;
+    }
+
+    protected invalidate(node: TreeNode) {
+        while (!this.dirty[node]) {
```
Comment (jack-williams) [inline packages/core/tree/test/bottomuptree.ts:16]: Should this be `if`?

**Review by jack-williams — APPROVED** [review]

**Pull request merged**